tx-lifecycle concept doc

gzhao408 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry: `clog add [section] [stanza] [message]`
- [ ] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

gzhao408 commented:
#3108

fedekunze commented:
shouldn't this be rebased to @gamarin2 branch ? #4426

alexanderbez commented:
Thanks @gzhao408!

@fedekunze probably? @gamarin2 what are your thoughts?

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,130 @@
+
+## State Changes
+During consensus, the validators came to agreement on not only which transactions but also the precise order of the transactions. However, apart from committing to this block in consensus, the ultimate goal is actually for nodes to commit to a new state generated by the transaction state changes. 
+In order to execute the transaction, the following ABCI function calls are made in order to communicate to the application what state changes need to be made. While nodes each run everything individually, since the messages' state transitions are deterministic and the order was finalized during consensus, this process yields a single, unambiguous result.
```
Where? How? 

Where: On each full-node (not necessarily validator nodes). 

How: Each full-node receives block candidates and validator signatures. The commit happens locally. All this is handled by Tendermint. It's important to clarify that, as readers might understand that full-node receive block AFTER they have been signed, whereas all this happens simultaneously (everybody, including non validator nodes, receive prevotes and precommits)

gamarin2 commented on docs/concepts/tx-lifecycle.md:
````diff
@@ -0,0 +1,130 @@
+		-----------------------
+```
+#### BeginBlock
+`BeginBlock` is run first, and mainly transmits important data such as block header and Byzantine Validators from the last round of consensus to be used during the next few steps. No transactions are handled here.
````
Link to `./app-anatomy.md#beginblocker-and-endblocker`

gamarin2 replied:
Also not technically part of tx lifecycle, but can remain

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,130 @@
+`BeginBlock` is run first, and mainly transmits important data such as block header and Byzantine Validators from the last round of consensus to be used during the next few steps. No transactions are handled here.
+
+#### DeliverTx
+The `DeliverTx` ABCI function defined in [`baseapp`](https://github.com/cosmos/cosmos-sdk/blob/8b1d75caa2099800ee9983e4a4528bcd00fec302/baseapp/baseapp.go
```
do not link to `baseapp` codebase here, but to `baseapp` concept doc. Only the concept doc of component X should directly reference code of component X

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,130 @@
+
+#### DeliverTx
+The `DeliverTx` ABCI function defined in [`baseapp`](https://github.com/cosmos/cosmos-sdk/blob/8b1d75caa2099800ee9983e4a4528bcd00fec302/baseapp/baseapp.go
+) does the bulk of the state change work: it is run for each transaction in the block in sequential order as committed to during consensus. Under the hood, `DeliverTx` is almost identical to `CheckTx` but calls the [`runTx`](https://github.com/cosmos/cosmos-sdk/blob/cec3065a365f03b86bc629ccb6b275ff5846fdeb/baseapp/baseapp.go#L757-L873) function in deliver mode instead of check mode. 
```
Same, link to `./baseapp.md#runtx` instead of the code directly

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,130 @@
+The `DeliverTx` ABCI function defined in [`baseapp`](https://github.com/cosmos/cosmos-sdk/blob/8b1d75caa2099800ee9983e4a4528bcd00fec302/baseapp/baseapp.go
+) does the bulk of the state change work: it is run for each transaction in the block in sequential order as committed to during consensus. Under the hood, `DeliverTx` is almost identical to `CheckTx` but calls the [`runTx`](https://github.com/cosmos/cosmos-sdk/blob/cec3065a365f03b86bc629ccb6b275ff5846fdeb/baseapp/baseapp.go#L757-L873) function in deliver mode instead of check mode. 
+
+The application utilizes both `AnteHandler` to check and `MsgHandler` to deliver, persisting changes in both `checkTxState` and `deliverTxState`, respectively. This second check happens because nodes may not have seen the same transactions in the same order during the Addition to Mempool stage and a malicious proposer may have included invalid transactions. 
```
`DeliverTx` should be explained more in-depth, as it constitutes the bulk of the tx lifecycle:

- Should outline the routing functionality of baseapp.
- Explain the role of modules
- Explain the lifecycle of the tx inside the module.

gzhao408 replied:
I think I should introduce modules earlier, such as include in the short definition section?

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,130 @@
+[`EndBlock`](https://github.com/cosmos/cosmos-sdk/blob/9036430f15c057db0430db6ec7c9072df9e92eb2/baseapp/baseapp.go#L875-L886) is always run at the end and is useful for automatic function calls or changing governance/validator parameters. No transactions are handled here.
+
+#### Commit
+The application's `Commit` method is run in order to finalize the state changes made by executing this block's transactions. A new state root should be sent back to serve as a merkle proof for the state change. Any application can inherit Baseapp's [`Commit`](https://github.com/cosmos/cosmos-sdk/blob/cec3065a365f03b86bc629ccb6b275ff5846fdeb/baseapp/baseapp.go#L888-L912) method; it synchronizes al ...
```
A bit confusing: sometimes it seems like we're following the lifecycle of a specific tx, and sometimes it feels like it's more of an abstract description of the lifecycle of txS. Need to choose one or the other

gamarin2 replied:
All the internal states (LIST), always good to list stuff again. Reiteration is how people learn

gamarin2 replied:
Re-explain why it's important that every internal state is reset after `Commit()`

gzhao408 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,123 @@
+4. **State Changes:** Transactions are delivered, creating state changes and expending gas. To commit state changes, internal state is updated and all copies are reset; the new state root is returned as proof.
+
+## Definition and Creation
+**[Transactions](https://github.com/cosmos/cosmos-sdk/blob/97d10210beb55ad4bd6722f7186a80bf7cb140e2/types/tx_msg.go#L36-L43)** are comprised of one or multiple **Messages** (link concept doc) and trigger state changes. The developer defines the specific messages to describe possible actions for the application by implementing the [`Msg`](https://github.com/cosmos/cosmos-sdk/blob/97d10210beb55ad4b ...
```
Ended up defaulting to plural "the nodes" and "the developers" and following one transaction `tx`

gzhao408 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,130 @@
+Each full node that receives a transaction performs local checks to filter out invalid transactions before they get included in a block. The transactions approved by a node are held in its [**Mempool**](https://github.com/tendermint/tendermint/blob/75ffa2bf1c7d5805460d941a75112e6a0a38c039/mempool/mempool.go) (memory pool unique to each node) pending approval from the rest of the network. Honest n ...
+
+### Stateless Checks
+Upon first receiving the transaction, an optional `PreCheckFunc` can be called to reject transactions that are clearly invalid as early as possible, such as ones exceeding the block size. Afterward, the next step is to unwrap the transaction into its message(s) and run each `validateBasic` function, which is simply a stateless sanity check (e.g. nonnegative numbers, nil strings, empty addresses). ...
```
hmm, there doesn't seem to be mention of `PreCheckFunc` in the Tendermint Mempool docs

gamarin2 replied:
Ok I thought you spotted it and that it had been recently added to the codebase. If it's not there, don't mention it :)

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,196 @@
+
+#### DeliverTx
+
+The `DeliverTx` ABCI function defined in [`baseapp`](./baseapp.md) does the bulk of the state change work: it is run for each transaction in the block in sequential order as committed to during consensus. Under the hood, `DeliverTx` is almost identical to `CheckTx` but calls the [`runTx`](./baseapp.md#runtx) function in deliver mode instead of check mode. Instead of using their `checkTxState` or  ...
```
Recurring issue:

You use "for each transaction" in the paragraph, and then take the viewpoint of a single tx. You have to choose between the two.

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,196 @@
+* **Initializations:** The `Context`, `Multistore`, and `GasMeter` are initialized.
+* **Checks:** The full-nodes call `validateBasicMsgs` and the `AnteHandler` again. This second check happens because they may not have seen the same transactions during the Addition to Mempool stage and a malicious proposer may have included invalid transactions.
+* **Route and Handler:** The extra step is to run `runMsgs` to fully execute each `Msg` within the transaction. Since `tx` may have messages from different modules, `baseapp` needs to know which module to find the appropriate Handler. Thus, the [`Route`](./msg-tx.md#route) function is called to retrieve the route name and find the `MsgHandler`.
+* **MsgHandler:** The `MsgHandler`, a step up from `AnteHandler`, is responsible for executing each message's actions and causes state changes to persist in `deliverTxState`. It is defined within a `Msg`'s module and writes to the appropriate stores within the module.
```
Simply called `Handler`

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,196 @@
+
+#### Commit
+
+The application's `Commit` method is run in order to finalize the state changes made by executing this block's transactions. A new state root should be sent back to serve as a merkle proof for the state change. Any application can inherit Baseapp's [`Commit`](./baseapp.md#commit) method; it syncs all the states by writing the `deliverTxState` into the application's internal state. Moving forward, ...
```
"this block" which block?

gamarin2 replied:
sent back where?

gamarin2 replied:
"Moving forward, `checkTxState`, `deliverTxState`, and `queryState` will all start afresh from the most recently committed state in order to be consistent and reflect the changes." when?

gzhao408 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,196 @@
+
+### Internal State
+
+**State** encompasses all relevant data in an application and is what the state machine stores. It includes data such as account balances, ownership of assets, and other relevant information that may change over time with executed transactions. State is stored in [Stores](./store/README.md).
```
Shortened the state section a bit into “types of checks”, hopefully flows better now

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,231 @@
+The `DeliverTx` ABCI function defined in [`baseapp`](./baseapp.md) does the bulk of the state change work: it is run for each transaction in the block in sequential order as committed to during consensus. Under the hood, `DeliverTx` is almost identical to `CheckTx` but calls the [`runTx`](./baseapp.md#runtx) function in deliver mode instead of check mode. Instead of using their `checkTxState` or  ...
+
+* **Decoding:** Since `DeliverTx` is an ABCI call, `Tx` is received in the encoded `byte []` form. Nodes first unmarshal the transaction, then call `runTx` in `runTxModeDeliver`, which does everything that `CheckTx` did and also executes and writes state changes.
+* **Checks:** The full-nodes call `validateBasicMsgs` and the `AnteHandler` again. This second check happens because they may not have seen the same transactions during the Addition to Mempool stage and a malicious proposer may have included invalid transactions.
```
Slight difference: `AnteHandler` will not check that `gas-prices > min-gas-prices`, since `min-gas-prices` is local to each node, and therefore can't be in consensus

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,231 @@
+* **Decoding:** Since `DeliverTx` is an ABCI call, `Tx` is received in the encoded `byte []` form. Nodes first unmarshal the transaction, then call `runTx` in `runTxModeDeliver`, which does everything that `CheckTx` did and also executes and writes state changes.
+* **Checks:** The full-nodes call `validateBasicMsgs` and the `AnteHandler` again. This second check happens because they may not have seen the same transactions during the Addition to Mempool stage and a malicious proposer may have included invalid transactions.
+* **Route and Handler:** While `CheckTx` would have exited, `DeliverTx` continues to run `runMsgs` to fully execute each `Msg` within the transaction. Since the transaction may have messages from different modules, `baseapp` needs to know which module to find the appropriate Handler. Thus, the [`Route`](./msg-tx.md#route) function is called to retrieve the route name and find the `Handler`.
+* **Handler:** The `Handler`, a step up from `AnteHandler`, is responsible for executing each message's actions and causes state changes to persist in `deliverTxState`. It is defined within a `Msg`'s module and writes to the appropriate stores within the module.
```
within a module

stores within the multistore, managed by the module

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,231 @@
+
+#### Commit
+
+After the hard work of executing `Tx` and all other transactions in its block, full-nodes need to finalize the state changes. A new state root is generated to serve as a merkle proof for the state change. Any application can inherit Baseapp's [`Commit`](./baseapp.md#commit) method; it syncs all the states by writing the `deliverTxState` into the application's internal state. As soon as the state  ...
```
remove "hard work", let's try to stay formal :p

gamarin2 replied:
`checkTxState` is reset to the latest state, but `DeliverState` is set to `nil` on commit, and initialized on `beginblock`

gamarin2 commented on docs/concepts/tx-msgs.md:
```diff
@@ -0,0 +1,28 @@
+# Messages and Transactions
```
Please move to a separate PR

gzhao408 commented on docs/concepts/store/README.md:
```diff
@@ -0,0 +1,64 @@
+# Multistore
```
oops accident haha

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,147 @@
+
+## State Changes
+
+During consensus, all validator-nodes come to agreement on not only whether `Tx` is to be included but also the precise order of the transactions in its block. However, apart from deciding upon this block in consensus, the ultimate goal is actually for full-nodes to commit to a new state generated by the state changes.
```
Can you rephrase this paragraph a bit? 

- It's confusing to single out `Tx` in this instance. I don't think the ordering of tx is important info in this context
- Don't use "However" in this instance
- Maybe a more factual description of the process would be better.

Factual description:

- Block proposal received by validator
- Begin Block -> DeliverTx -> EndBlock
- Prevote
- Precommit
- Commit

gzhao408 replied:
mm, i was confused on the ordering. I changed quite a bit and added a Commit subsection after state changes.

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,147 @@
+The `DeliverTx` ABCI function defined in [`baseapp`](./baseapp.md) does the bulk of the state change work: it is run for each transaction in the block in sequential order as committed to during consensus. Under the hood, `DeliverTx` is almost identical to `CheckTx` but calls the [`runTx`](./baseapp.md#runtx) function in deliver mode instead of check mode. Instead of using their `checkState` or `q ...
+
+* **Decoding:** Since `DeliverTx` is an ABCI call, `Tx` is received in the encoded `[]byte` form. Nodes first unmarshal the transaction, then call `runTx` in `runTxModeDeliver`, which does everything that `CheckTx` did but also executes and writes state changes.
+* **Checks:** Full-nodes call `validateBasicMsgs` and the `AnteHandler` again. This second check happens because they may not have seen the same transactions during the Addition to Mempool stage and a malicious proposer may have included invalid ones.
```
I don't think this comment was addressed "Slight difference: AnteHandler will not check that gas-prices > min-gas-prices, since min-gas-prices is local to each node, and therefore can't be in consensus"

gzhao408 replied:
ohhh ok I misunderstood before.

gzhao408 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,147 @@
+
+The `Context` used to keep track of important data while `AnteHandler` is executing `Tx` keeps a `GasMeter` which tracks how much gas has been used. The user-provided amount for gas is known as the value `GasWanted`. If `GasConsumed`, the amount of gas consumed so far, ever exceeds `GasWanted`, execution stops. Otherwise, `CheckTx` sets `GasUsed` equal to `GasConsumed` and returns it in the resul ...
+
+If at any point during the checking stage `Tx` fails a check, it is discarded and the transaction lifecycle ends here. Otherwise, if it passes `CheckTx` successfully, the default protocol is to relay it to peer nodes and add it to the Mempool so that `Tx` becomes a candidate to be included in the next block. Note that even if `Tx` passes all checks at this stage, it is still possible to be found  ...
```
got it. summarized to make it shorter, since i can imagine you saying "it breaks up the flow" of the lifecycle doc :P

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+
+## Addition to Mempool
+
+Each full node that receives `Tx` performs local checks to ensure it is not invalid. If the transaction passes the checks, it is held in the nodes' [**Mempool**](https://tendermint.com/docs/tendermint-core/mempool.html#mempool)s (memory pools of transactions unique to each node) pending inclusion in a block. Honest nodes will discard `Tx` if it is found to be invalid. Prior to consensus, nodes co ...
```
mention `CheckTx` in this paragraph

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+
+### Types of Checks
+
+The full-node performs stateless, then stateful checks on `Tx`, with the goal to identify and reject an invalid transaction as early on as possible to avoid wasted computation. ***Stateless*** checks do not require nodes to access state - light clients or offline nodes can do them - and are thus less computationally expensive. Stateless checks include making sure addresses are not empty, enforcin ...
```
"the full-node perform ... *during CheckTx*"

You can then say that the same checks are performed during `DeliverTx` in the `DeliverTx` paragraph

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+
+### ValidateBasic
+
+[Messages](./tx-msgs.md#messages) are extracted from `Tx` and [`ValidateBasic`](./msg-tx.md#validatebasic), a function defined for every message, is run for each one. It should include basic stateless sanity checks. For example, if the message is to send coins from one address to another, `ValidateBasic` likely checks for nonempty addresses and a nonnegative coin amount, but does not require know ...
```
Specify that `ValidateBasic` needs to be implemented by the module developer

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+
+The [`AnteHandler`](./baseapp.md#antehandler), which is technically optional but should be defined for each application, is run. A deep copy of the internal state, `checkState`, is made and the defined `AnteHandler` performs limited checks specified for the transaction type. Using a copy allows the handler to do stateful checks for `Tx` without modifying the last committed state, and revert back  ...
+
+For example, the `auth` module `AnteHandler` checks and increments sequence numbers, checks signatures and account numbers, and deducts fees from the first signer of the transaction - all state changes are made using the `checkState`. 
```
Link to `auth` module

gzhao408 replied:
used the auth spec: https://github.com/cosmos/cosmos-sdk/tree/master/docs/spec/auth

gamarin2 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+The `DeliverTx` ABCI function defined in [`baseapp`](./baseapp.md) does the bulk of the state change work: it is run for each transaction in the block in sequential order as committed to during consensus. Under the hood, `DeliverTx` is almost identical to `CheckTx` but calls the [`runTx`](./baseapp.md#runtx-and-runmsgs) function in deliver mode instead of check mode. Instead of using their `check ...
+
+* **Decoding:** Since `DeliverTx` is an ABCI call, `Tx` is received in the encoded `[]byte` form. Nodes first unmarshal the transaction, then call `runTx` in `runTxModeDeliver`, which is very similar to `CheckTx` but also executes and writes state changes.
+* **Checks:** Full-nodes call `validateBasicMsgs` and the `AnteHandler` again. This second check happens because they may not have seen the same transactions during the Addition to Mempool stage and a malicious proposer may have included invalid ones. One difference here is that the `AnteHandler` will not compare `gas-prices` to the node's `min-gas-prices`since that value is local to each node.
```
Good, but add little explanation of why local values cannot be taken into account during `DeliverTx` (because state-changes must be deterministic, i.e. everyone must have the same state at the end of a block)

hschoenburg commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+The final step is for validator-nodes participating in consensus to commit the block and state changes. Validator-nodes perform the previous step of executing state changes in order to validate the transactions, then sign the block to confirm and vote for them. Full-nodes that are not validators do not participate in consensus - i.e. they cannot vote - but listen for votes to understand whether o ...
+
+#### Prevote and Precommit
+
```
hmm maybe this is overcrowding the document to describe Tendermint BFT as well? This doc could just focus on how the SDK handles transactions

gzhao408 replied:
I'll condense/generic-ify it

gzhao408 commented on docs/concepts/tx-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+
+### Transaction Creation
+
+The transaction `Tx` is created by the user inputting a command in the following format from the [command-line]((./interfaces.md#cli)), providing the type of transaction in `[command]`, arguments in `[args]`, and configurations such as gas prices in `[flags]`:
```
Thank you Hans!

jackzampolin commented:
All this needs is a rebase @gzhao408 then I will merge!

fedekunze commented:
@gzhao408 mind resolving the comments that have already been addressed so we can move forward merging this PR?

fedekunze commented:
Why are there so many files on the diff?

gzhao408 commented:
> Why are there so many files on the diff?

My apologies, I didn't realize Gautier had moved `tx-lifecycle.md` and deleted the concepts folder to reorganize docs, so in my git-confused-ness I had added it back in the merging. Just made a new PR to undo the changes: #4739
